File: podman-script/pr.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"gopkg.in/yaml.v2"
	"github.com/jessevdk/go-flags"
	"os/exec"
)

// Configuration struct to hold the values from the YAML
type Config struct {
	LocalPath       string `yaml:"local_path"`
	ContainerVolume string `yaml:"container_volume"`
	YamlFile        string `yaml:"yaml_file"`
	RegistryImage   string `yaml:"registry_image"`
}

var configFile = "/Users/geoagriogiannis/Documents/scripts/pr.yaml"

// Options for command-line arguments
type Options struct {
	LocalPath       string `short:"l" long:"local_path" description:"Local directory path"`
	ContainerVolume string `short:"v" long:"container_volume" description:"Container volume path"`
	YamlFile        string `short:"y" long:"yaml_file" description:"Path to the YAML definition file"`
	RegistryImage   string `short:"r" long:"registry" description:"Registry image (optional)"`
	Help            bool   `short:"h" long:"help" description:"Show this help message"`
}

// Function to print the usage information
func printUsage() {
	fmt.Println("Usage: pr.go [options]")
	fmt.Println("\nOptions:")
	fmt.Println("  -l, --local_path PATH           Local directory path")
	fmt.Println("  -v, --container_volume PATH     Container volume path")
	fmt.Println("  -y, --yaml_file PATH            Path to the YAML definition file")
	fmt.Println("  -r, --registry IMAGE            Registry image (optional)")
	fmt.Println("  -h, --help                      Show this help message")
	fmt.Printf("\nThe default values are loaded from the YAML file located at:\n%s\n", configFile)
}

func main() {
	// Parse command-line options
	var opts Options
	_, err := flags.Parse(&opts)
	if err != nil {
		fmt.Println("Error parsing options:", err)
		return
	}

	// Check if --help is requested, if yes, print usage and exit
	if opts.Help {
		printUsage()
		return
	}

	// Load the configuration file
	configFilePath := configFile
	config := Config{}
	data, err := os.ReadFile(configFilePath)
	if err != nil {
		fmt.Println("Error loading YAML configuration file:", err)
		return
	}

	// Unmarshal YAML into config struct
	err = yaml.Unmarshal(data, &config)
	if err != nil {
		fmt.Println("Error unmarshaling YAML:", err)
		return
	}

	// Set the values from command-line options if provided, otherwise use the config values
	if opts.LocalPath == "" {
		opts.LocalPath = config.LocalPath
	}
	if opts.ContainerVolume == "" {
		opts.ContainerVolume = config.ContainerVolume
	}
	if opts.YamlFile == "" {
		opts.YamlFile = config.YamlFile
	}
	if opts.RegistryImage == "" {
		opts.RegistryImage = config.RegistryImage
	}

	// Validate inputs
	if opts.LocalPath == "" {
		fmt.Println("Error: Local path is required.")
		return
	}
	if opts.ContainerVolume == "" {
		fmt.Println("Error: Container volume is required.")
		return
	}
	if opts.YamlFile == "" {
		fmt.Println("Error: YAML file is required.")
		return
	}
	if opts.RegistryImage == "" {
		fmt.Println("Error: Registry image is required.")
		return
	}

	// Construct the podman command
	command := fmt.Sprintf("podman run --rm -it --privileged -v %s:%s %s build --definition-file %s", opts.LocalPath, opts.ContainerVolume, opts.RegistryImage, opts.YamlFile)

	// Ask for confirmation to run the command
	fmt.Println("\nThe following command will be executed:")
	fmt.Println(command)

	fmt.Print("Execute this command? (yes/no): ")
	reader := bufio.NewReader(os.Stdin)
	confirmation, _ := reader.ReadString('\n')
	confirmation = strings.TrimSpace(confirmation)
	if confirmation == "" {
		confirmation = "yes"
	}

	// If the user confirms, run the command
	if confirmation == "yes" {
		// Run the podman command
		err := runCommand(command)
		if err != nil {
			fmt.Println("Error executing the command:", err)
		}
	} else {
		fmt.Println("Command execution aborted.")
	}
}

// Function to actually execute the command
func runCommand(command string) error {
	cmd := exec.Command("bash", "-c", command)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

File: podman-script/pr.rb
#!/usr/bin/env ruby
require 'optparse'
require 'yaml'

# Define the path to the configuration file
CONFIG_FILE = File.expand_path('~/Documents/scripts/pr.yaml')

# OptionParser to handle command-line options and arguments
options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: pr.rb [options]"

  opts.on("-l", "--local_path PATH", "Local directory path") do |v|
    options[:local_path] = v
  end

  opts.on("-v", "--container_volume PATH", "Container volume path") do |v|
    options[:container_volume] = v
  end

  opts.on("-y", "--yaml_file PATH", "Path to the YAML definition file") do |v|
    options[:yaml_file] = v
  end

  opts.on("-r", "--registry IMAGE", "Registry image (optional)") do |v|
    options[:registry_image] = v
  end

  opts.on("-h", "--help", "Show this help message") do
    # Print the usage directly here and exit
    puts <<-USAGE
Usage: pr.rb [options]

Options:
  -l, --local_path PATH           Local directory path
  -v, --container_volume PATH     Container volume path
  -y, --yaml_file PATH            Path to the YAML definition file
  -r, --registry IMAGE            Registry image (optional)
  -h, --help                      Show this help message

The default values are loaded from the YAML file located at:
#{CONFIG_FILE}
    USAGE
    exit 0 # Exit after printing help
  end
end.parse!

# Load the configuration file
config = YAML.load_file(CONFIG_FILE)

# Set default values from the config file if not provided via options
local_path = options[:local_path] || config['local_path'] || ''
container_volume = options[:container_volume] || config['container_volume'] || ''
yaml_file = options[:yaml_file] || config['yaml_file'] || ''
registry_image = options[:registry_image] || config['registry_image'] || ''

# Validate the input paths
if local_path.empty?
  puts "Error: Local path is required."
  exit 1
end

# if container_volume.empty?
#   puts "Error: Container volume is required."
#   exit 1
# end

if yaml_file.empty?
  puts "Error: YAML file is required."
  exit 1
end

if registry_image.empty?
  puts "Error: Registry image is required."
  exit 1
end

# Print out the configuration for debugging
puts "Loaded configuration from #{CONFIG_FILE}:"
puts YAML.dump(config)

# Construct the podman command
command = "podman run --rm -it --privileged -v #{local_path}:/eib #{registry_image} build --definition-file #{yaml_file}" # container image hardcoded for eib

# Print the command for debugging
puts "Executing this command:"
puts command

# Ask for confirmation to run the command
print "Execute this command? (yes/no): "
confirmation = gets.chomp.downcase
# If input is empty, treat it as 'yes'
confirmation = 'yes' if confirmation.empty?

if confirmation == 'yes'
  # Run the podman command
  system(command)  # This will actually execute the command
else
  puts "Command execution aborted."
end